### ISSUE TITLE
Zoom resets on mouse move

### ISSUE THREAD

**Issue by ilyapopov** [hanatos/vkdt#20]
When I open an image, I can zoom in and out with the mouse wheel. But once I move the mouse even one pixel, zoom resets back to the initial size. 

Ubuntu 21.04, Xorg session, GCC 10.3.0, GLFW 3.3.2-1

**Comment by hanatos** [comment]
interesting, i cannot reproduce. maybe the window resizes because of the window manager or something? maybe an erroneous middle mouse button event triggers? there aren't that many places in the code to reset the view.

any hints for me to try and reproduce? any particular processing graph or image format? window manager? full screen? opening a folder or a single image?

**Comment by ilyapopov** [comment]
I opened a folder first. Than from folder I opened one image. I used default pipeline, did not change any settings. I am using Ubuntu 21.04 with X.org session. The window was maximized but not full screen. I will try to provide more detail later.

**Comment by ilyapopov** [comment]
I did a little printf debugging, and reset happens in `style_to_state` function
https://github.com/hanatos/vkdt/blob/2694a002e166f07c1d47b36905cf2b4b41fbb2b6/src/gui/gui.c#L19

which is called from `dt_gui_recreate_swapchain` function,
https://github.com/hanatos/vkdt/blob/2694a002e166f07c1d47b36905cf2b4b41fbb2b6/src/gui/gui.c#L192

 which is called directly from main here:

https://github.com/hanatos/vkdt/blob/2694a002e166f07c1d47b36905cf2b4b41fbb2b6/src/gui/main.c#L220

This happens on every mouse move. It does not happen then mouse does not move.

**Comment by ilyapopov** [comment]
And for the record, these are the messages I am getting in terminal on startup:

```
[ERR] module burst has no connectors!
[gui] monitor [0] DP-2-2 at 0 0
[gui] vk extension required by GLFW:
[gui]   VK_KHR_surface
[gui]   VK_KHR_xcb_surface
MESA-INTEL: warning: Performance support disabled, consider sysctl dev.i915.perf_stream_paranoid=0

[ERR] device does not support requested feature samplerAnisotropy, trying anyways
[ERR] device does not support requested feature shaderStorageImageExtendedFormats, trying anyways
[ERR] device does not support requested feature shaderUniformBufferArrayDynamicIndexing, trying anyways
[ERR] device does not support requested feature shaderSampledImageArrayDynamicIndexing, trying anyways
[ERR] device does not support requested feature shaderStorageBufferArrayDynamicIndexing, trying anyways
[ERR] device does not support requested feature shaderStorageImageArrayDynamicIndexing, trying anyways
[ERR] device does not support requested feature inheritedQueries, trying anyways
Style to state
[gui] no display profile file display.DP-2-2, using sRGB!
[gui] no display profile file display.DP-2-2, using sRGB!
key_callback
Style to state
[ERR] could not open directory '(null)'!
[ERR] [thm] no images in list!
```

So maybe it's Intel GPU that causes the problem?

**Comment by hanatos** [comment]
thanks for digging up all that information!

yeah going into the else branch for

`if(dt_gui_render() == VK_SUCCESS)`

does not sound too reassuring.

can you run with `./vkdt -d all` or so to see whether there are any
vulkan/validationlayer complaints on the way? it's possible that this
triggers for super long running renders, i think i enforced a timeout that
will not return VK_SUCCESS then.

**Comment by ilyapopov** [comment]
when started as `./vkdt -d all` it produces the following output when moving the mouse:

```
[qvk] available surface formats:
[qvk] B8G8R8A8_SRGB
[qvk] B8G8R8A8_UNORM
[qvk] colour space: 0
[perf] ui time 0.000s
[perf] ui time 0.000s
[perf] ui time 0.000s
[perf] ui time 0.000s
[perf] ui time 0.000s
[qvk] error VK_TIMEOUT executing vkAcquireNextImageKHR(qvk.device, qvk.swap_chain, 1ul<<20, image_acquired_semaphore, VK_NULL_HANDLE, &vkdt.frame_index)!
```

**Comment by ilyapopov** [comment]
On startup is prints this:

```
[ERR] module burst has no connectors!
[gui] monitor [0] DP-2-2 at 0 0
[gui] vk extension required by GLFW:
[gui]   VK_KHR_surface
[gui]   VK_KHR_xcb_surface
MESA-INTEL: warning: Performance support disabled, consider sysctl dev.i915.perf_stream_paranoid=0

[qvk] dev 0: vendorid 0x8086
[qvk] dev 0: Intel(R) HD Graphics 520 (SKL GT2)
[qvk] max number of allocations -1
[qvk] max image allocation size 16384 x 16384
[qvk] max uniform buffer range 134217728
[qvk] dev 1: vendorid 0x10005
[qvk] dev 1: llvmpipe (LLVM 11.0.1, 256 bits)
[qvk] max number of allocations 4096
[qvk] max image allocation size 16384 x 16384
[qvk] max uniform buffer range 65536
[ERR] device does not support requested feature samplerAnisotropy, trying anyways
[ERR] device does not support requested feature shaderStorageImageExtendedFormats, trying anyways
[ERR] device does not support requested feature shaderUniformBufferArrayDynamicIndexing, trying anyways
[ERR] device does not support requested feature shaderSampledImageArrayDynamicIndexing, trying anyways
[ERR] device does not support requested feature shaderStorageBufferArrayDynamicIndexing, trying anyways
[ERR] device does not support requested feature shaderStorageImageArrayDynamicIndexing, trying anyways
[ERR] device does not support requested feature inheritedQueries, trying anyways
[qvk] picked device 0 without ray tracing support
[qvk] num queue families: 1
[qvk] validation layer: Loading layer library libVkLayer_MESA_device_select.so
[qvk] available surface formats:
[qvk] B8G8R8A8_SRGB
[qvk] B8G8R8A8_UNORM
[qvk] colour space: 0
[gui] no display profile file display.DP-2-2, using sRGB!
[gui] no display profile file display.DP-2-2, using sRGB!
[qvk] available surface formats:
[qvk] B8G8R8A8_SRGB
[qvk] B8G8R8A8_UNORM
[qvk] colour space: 0
[db] allocating 1024.0 MB for thumbnails
[mem] images : peak rss 0.00390625 MB vmsize 0.00390625 MB
[mem] staging: peak rss 0.000244141 MB vmsize 0.000244141 MB
[perf] record cmd buf:	   4.909 ms
[perf] i-bc1_main:	   0.035 ms
[perf] total align:	   0.000 ms
[perf] total time:	   0.035 ms
[perf] [thm] ran graph in 281ms
[mem] images : peak rss 0.00390625 MB vmsize 0.00390625 MB
[mem] staging: peak rss 0.000244141 MB vmsize 0.000244141 MB
[perf] record cmd buf:	   0.227 ms
[perf] i-bc1_main:	   0.032 ms
[perf] total align:	   0.000 ms
[perf] total time:	   0.032 ms
[perf] [thm] ran graph in   0ms
[ERR] could not open directory '(null)'!
[ERR] [thm] no images in list!
[qvk] available surface formats:
[qvk] B8G8R8A8_SRGB
[qvk] B8G8R8A8_UNORM
[qvk] colour space: 0
[perf] ui time 0.000s
[perf] ui time 0.001s
[perf] ui time 0.000s
[perf] ui time 0.001s
[perf] ui time 0.001s
[perf] ui time 0.000s
[perf] ui time 0.001s
```

**Comment by hanatos** [comment]
right. thanks for confirming. it indeed seems to be the timeout. if you
don't mind the wait and want to stick to this hardware you can probably
increase the timeout value in

src/pipe/graph.c: lines 2233 and 2312: any occurrence of
`QVKR(vkWaitForFences(qvk.device, 1, &graph->command_fence, VK_TRUE,
1ul<<40));`

by increasing the number in this last parameter (make it 1ul<<50?).

i seem to remember running simple images on a HD 520 just fine though.
maybe the hint:

`MESA-INTEL: warning: Performance support disabled, consider sysctl
dev.i915.perf_stream_paranoid=0`

actually means something here?

**Comment by ilyapopov** [comment]
I would love to have a better GPU for photographic tasks, but GPUs are not easy to come by nowadays. 

Neither sysctl nor changing timeout to 1ul<<50 has helped.

Also, if that helps, the ouput when opening an image is:

```
[rawspeed] load /home/ipopov/Pictures/2021/06-19/Z06_5591.NEF in 291ms
[mem] images : peak rss 342.609 MB vmsize 362.609 MB
[mem] staging: peak rss 47.1301 MB vmsize 47.1301 MB
[perf] record cmd buf:	 312.686 ms
[perf] i-raw_main:	  13.955 ms
[perf] denoise_noop:	  45.720 ms
[perf] hilite_half:	  15.838 ms
[perf] hilite_reduce:	  15.513 ms
[perf] hilite_reduce:	   3.735 ms
[perf] hilite_reduce:	   0.967 ms
[perf] hilite_reduce:	   0.248 ms
[perf] hilite_reduce:	   0.066 ms
[perf] hilite_reduce:	   0.023 ms
[perf] hilite_reduce:	   0.018 ms
[perf] hilite_reduce:	   0.017 ms
[perf] hilite_assemble:	   0.016 ms
[perf] hilite_assemble:	   0.009 ms
[perf] hilite_assemble:	   0.021 ms
[perf] hilite_assemble:	   0.066 ms
[perf] hilite_assemble:	   0.243 ms
[perf] hilite_assemble:	   1.008 ms
[perf] hilite_assemble:	   3.994 ms
[perf] hilite_assemble:	  15.938 ms
[perf] hilite_doub:	  14.442 ms
[perf] demosaic_halfsize:	   6.819 ms
[perf] shared_resample:	   1.454 ms
[perf] crop_main:	   0.160 ms
[perf] colour_main:	   0.209 ms
[perf] filmcurv_main:	   0.189 ms
[perf] llap_curve:	   0.321 ms
[perf] llap_reduce:	   0.320 ms
[perf] llap_reduce:	   0.084 ms
[perf] llap_reduce:	   0.026 ms
[perf] llap_reduce:	   0.014 ms
[perf] llap_reduce:	   0.011 ms
[perf] llap_reduce:	   0.009 ms
[perf] llap_reduce:	   0.009 ms
[perf] llap_reduce:	   0.008 ms
[perf] llap_assemble:	   0.035 ms
[perf] llap_assemble:	   0.019 ms
[perf] llap_assemble:	   0.012 ms
[perf] llap_assemble:	   0.014 ms
[perf] llap_assemble:	   0.020 ms
[perf] llap_assemble:	   0.046 ms
[perf] llap_assemble:	   0.136 ms
[perf] llap_assemble:	   0.513 ms
[perf] llap_colour:	   0.095 ms
[perf] f2srgb_main:	   0.057 ms
[perf] total align:	   1.454 ms
[perf] total time:	 142.607 ms
[perf] [thm] ran graph in 499ms

```

**Comment by hanatos** [comment]
oh my! these numbers are terrible! 45ms for noop! i'll have to dig up my
old laptop and see whether i can produce numbers that are *this* bad.

until then there's a half-assed feature to process low res for display. in
the "esoteric" tab there's a "lod" slider. if you set this to "2" it will
only process at the resolution of your monitor, not the full res of the
input (increase even more for even coarser previews). if the gui is
dysfunctional to a point where you can't even do this by hand, you can edit
`~/.config/vkdt/config.rc` and make sure it contains a line like
`intgui/lod:2`.

be warned that i almost never use this feature, so i'll have to assume it's
broken somewhere. in particular the hand drawn masks in the `draw` module
as well as the blur radii in guided filters may produce unexpected results
(i should fix/double check this).

i'm still confused a bit by the timeout, which should be longer than the
~0.5s the pipeline seems to report there. on the other hand these timings
are for thumbnail generation (see last line `[thm]`), the darkroom mode may
be even worse.

**Comment by hanatos** [comment]
closing this a kinda resolved, i suppose the only action item on my side would be auto detect lod and make sure it actually works or something.

**Comment by ilyapopov** [comment]
A little update from my side.

Since last time, I have got hold of a better machine, with Ryzen 5700G APU (also an integrated GPU, but much better than that Intel), and the issue was exactly the same. The application runs without visible delays, but there was a timeout in the console and zoom resetting on mouse move. I have tried all your suggestions and nothing helped. LOD did not help either.

I am happy to report that the issue is finally gone with commit db07604fe59c8da491f3fbc3fb569d801d7fa0a9

And it works smooth enough, without visible delays. 

There is however, an problem with thumbnails, I will create a separate issue for it.
